### PR TITLE
Log-316: Elasticsearch Operator Telemetry

### PR DESCRIPTION
**Pull request by sasagarw**
### Description
This PR:
- Exposes custom metrics which can be scraped by monitoring prometheus.
- This can be queried via the OCP dashboard (Monitoring -> Metrics).

/cc @periklis 
/assign @ewolinetz 

### Links

- JIRA: https://issues.redhat.com/browse/LOG-316

### CODE REVIEW

File: internal/metrics/metrics.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+)
+
+var (
+	nodeRestartReason = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "eo_elasticsearch_cr_node_restart_reason",
+			Help: "Node restarted due to Cert Restart or Rolling Restart (1 = Cert Restart, 2 = Rolling Restart, 3 = Scheduled).",
+		})
+
+	rollingRestartsCounter = prometheus.NewCounter(
```
Comment (periklis) [inline internal/metrics/metrics.go:15]: As per talk:
- Counters should be expressed as `****_total`
- Fold the `nodeRestartReason` metric into this metric using a label `reason={cert_restart, rolling_restart, scheduled_restart`

Consider please best practices: https://prometheus.io/docs/practices/naming/

File: internal/metrics/metrics.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+		prometheus.CounterOpts{
+			Name: "eo_elasticsearch_cr_rolling_restart_counter",
+			Help: "Number of times the nodes restarted due to Rolling Restart.",
+		})
+
+	esClusterManagementState = prometheus.NewGauge(
```
Comment (periklis) [inline internal/metrics/metrics.go:21]: As per talk:
- Managed/Unmanaged should be a label.

File: internal/metrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+//Sets the metric value to "3" when the node restarts due to cert redeploy or rolling update is scheduled for.
+func UpdateMetricNodeRestartScheduled() {
+	nodeRestartReason.Set(3)
+}
+
+//Increment the metric value when the node restarts due to rolling update.
+func IncrementMetricNodeRestartRolling() {
+	rollingRestartsCounter.Inc()
```
Comment (periklis) [inline internal/metrics/metrics.go:62]: `rollingRestartsCounter.Inc()` -> `rollingRestartsCounter.WithLabels(label = rolling).Inc()`

File: internal/metrics/metrics.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+//Sets the metric value to "1" when the ES Cluster Management State is Managed.
+func UpdateMetricESClusterManagementStateManaged() {
+	esClusterManagementState.Set(1)
```
Comment (periklis) [inline internal/metrics/metrics.go:67]: - Think of this case `esClusterManagementState.WithLabels(state = Managed).Set(10)`?
- Consider renaming to `SetEsClusterManagementStateManaged(int)`

File: internal/metrics/metrics.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+//This function registers the custom metrics to the kubernetes controller-runtime default metrics.
+func RegisterCustomMetrics() {
+	for _, metric := range metricList {
+		metrics.Registry.MustRegister(metric)
+	}
+
+	//Set initial value for metrics as "0".
+	nodeRestartReason.Set(0)
+	esClusterManagementState.Set(0)
```
Comment (periklis) [inline internal/metrics/metrics.go:42]: Is this needed, as metrics get default initialized

**Review by periklis — CHANGES_REQUESTED** [review]
Awesome work! As we talked online, please consider the comments on `metrics.go`.

File: internal/metrics/metrics.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+		"state": ManagedState,
+	}).Set(float64(number))
+}
+
+//Sets the metric value to "n" when the ES Cluster Management State is Unmanaged.
+func SetEsClusterManagementStateUnmanaged(number int) {
+	esClusterManagementState.With(prometheus.Labels{
+		"state": UnmanagedState,
+	}).Set(float64(number))
+}
```
Comment (periklis) [inline internal/metrics/metrics.go:75]: Reflecting a little bit on the [prometheus/client-go gauge interface](https://github.com/prometheus/client_golang/blob/master/prometheus/gauge.go#L32), it looks to me that using `Set` is the wrong approach. We want do use `Add` for adding/removing clusters from each metric, i.e.:
- Cluster Managed -> Unmanaged: `Managed.Add(-1)` and `Unmanaged.Add(1)`
- Cluster Unmanager -> Managed: `Managed.Add(1)` and `Unmanaged.Add(-1)`

WDYT? (cc @blockloop)

File: internal/metrics/metrics.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+const (
+	CertRestart 		string = "cert_restart"
+	RollingRestart 		string = "rolling_restart"
+	ScheduledRestart 	string = "scheduled_restart"
```
Comment (periklis) [inline internal/metrics/metrics.go:11]: @ewolinetz Could you please elaborate which restart case is scheduled only, but neither cert-restart nor rolling-restart?

**Review by periklis — COMMENTED** [review]
Great progress! Thank you so much :rocket: 

I suggested some small improvements regarding prometheus/client-go usage and put some ask for confirmation to @ewolinetz and @blockloop .

File: internal/metrics/metrics.go (hunk 6 above)
Comment (sasagarw) [reply internal/metrics/metrics.go:75]: Indeed you are correct, `Set` will assign a constant value whatever we pass as param. Rather what we want here is to increment and decrement values whenever the cluster changes it's management set. `Add` helps us to achieve this. But the values might go negative as well. For example, initially both will be 0, when  we create a CR with `managed` state then `manage` will become 1 whereas `unmanaged` will become -1

File: internal/metrics/metrics.go (hunk 7 above)
Comment (sasagarw) [reply internal/metrics/metrics.go:11]: According to what I understood, `scheduled_restart` isn't something different. I guess when `cert_restart` or `rolling_restart` is unsuccessful due to some reason then we schedule them for sometime in future. Is this what it means @ewolinetz ?

File: controllers/logging/elasticsearch_controller.go [hunk 8]
```diff
@@ -49,8 +51,14 @@ func (r *ElasticsearchReconciler) Reconcile(request ctrl.Request) (ctrl.Result,
...
 
 	if cluster.Spec.ManagementState == loggingv1.ManagementStateUnmanaged {
+		// Cluster state changes from Managed -> Unmanaged, so set "unmanaged" as 1 and set "managed" as 0.
+		metrics.SetEsClusterManagementStateManaged(0)
+		metrics.SetEsClusterManagementStateUnmanaged(1)
 		return ctrl.Result{}, nil
 	}
+	// Cluster state changes from Unmanaged -> Managed, so set "managed" as 1 and set "unmanaged" as 0.
+	metrics.SetEsClusterManagementStateManaged(1)
+	metrics.SetEsClusterManagementStateUnmanaged(0)
```
Comment (periklis) [inline controllers/logging/elasticsearch_controller.go:61]: nit. You could have hidden this numbers behind the helper functions, e.g. `metrics.SetEsClusterManagementStateManaged` is doing the `SetManaged(1)` and `SetUnmanaged(0)` in the background.

**Review by blockloop — APPROVED** [review]
/lgtm

File: internal/k8shandler/reconciler.go [hunk 9]
```diff
@@ -72,6 +73,7 @@ func SecretReconcile(requestCluster *elasticsearchv1.Elasticsearch, requestClien
 			if err := requestClient.Status().Update(context.TODO(), cluster); err != nil {
 				return err
 			}
+			metrics.IncrementRestartCounterCert()
```
Comment (ewolinetz) [inline internal/k8shandler/reconciler.go:76]: i think this is going to increment this twice... if we have it here in this reconciler but then also in the code where we perform the cert restart

File: internal/metrics/metrics.go (hunk 7 above)
Comment (ewolinetz) [reply internal/metrics/metrics.go:11]: `cert_restart` means we are scheduled to restart for the certificate.
`scheduled_restart` means it is scheduled to be updated after... i'm not sure what `rolling_restart` is covering here.

We currently just have two different scheduled restart cases

File: internal/k8shandler/reconciler.go (hunk 9 above)
Comment (sasagarw) [reply internal/k8shandler/reconciler.go:76]: Corrected

File: internal/metrics/metrics.go (hunk 7 above)
Comment (periklis) [reply internal/metrics/metrics.go:11]: @ewolinetz What is your suggestion here? Drop the label value `rolling_restart`? If yes, then the AC is broken too?
```
Expose Operator's current action (e.g. rolling restart, full restart, idle, managed, unmanaged) – Cluster Condition
```

Comment (ewolinetz) [reply internal/metrics/metrics.go:11]: I was strictly answering the question asked. I hadn't looked at the AC in relation to what was being done here.
Rolling_restart would be that it currently is in the process of performing a rolling restart.

**Comment by sasagarw** [comment]
@periklis @ewolinetz can I get labels on this now?

**Comment by periklis** [comment]
> @periklis @ewolinetz can I get labels on this now?

It is still not clear to me if we @ewolinetz suggests to drop the `rolling_restart` label or not. Otherwise looks good to me.

**Comment by sasagarw** [comment]
@periklis Can I get lgtm now?

**Pull request merged**